File: unity-ads/src/test/java/com/unity3d/services/ads/gmascar/managers/BiddingManagerFactoryTest.java
package com.unity3d.services.ads.gmascar.managers;

import static org.hamcrest.CoreMatchers.instanceOf;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

import com.unity3d.ads.IUnityAdsTokenListener;
import com.unity3d.services.ads.gmascar.GMA;
import com.unity3d.services.core.configuration.IExperiments;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.powermock.api.mockito.PowerMockito;
import org.powermock.core.classloader.annotations.PrepareForTest;
import org.powermock.modules.junit4.PowerMockRunner;


@RunWith(PowerMockRunner.class)
@PrepareForTest({GMA.class})
public class BiddingManagerFactoryTest {

	BiddingManagerFactory biddingManagerFactory = BiddingManagerFactory.getInstance();
	IExperiments experiments = mock(IExperiments.class);
	GMA gma = mock(GMA.class);

	@Before
	public void setup() throws Exception {
		PowerMockito.whenNew(GMA.class).withAnyArguments().thenReturn(gma);
		when(gma.hasSCARBiddingSupport()).thenReturn(true);
	}

	@Test
	public void createManagerNoExperiment() {
		BiddingBaseManager manager = biddingManagerFactory.createManager(new IUnityAdsTokenListener() {
			@Override
			public void onUnityAdsTokenReady(String token) {

			}
		}, null);
		assertThat(manager, instanceOf(BiddingEagerManager.class));
	}

	@Test
	public void createManagerNoListener() {
		BiddingBaseManager manager = biddingManagerFactory
			.createManager(null, experiments);
		assertThat(manager, instanceOf(BiddingEagerManager.class));
	}

	@Test
	public void createManagerExperimentEager() {
		when(experiments.getScarBiddingManager()).thenReturn("eag");

		BiddingBaseManager manager = biddingManagerFactory
			.createManager(mock(IUnityAdsTokenListener.class), experiments);
		assertThat(manager, instanceOf(BiddingEagerManager.class));
	}

	@Test
	public void createManagerExperimentLazy() {
		when(experiments.getScarBiddingManager()).thenReturn("laz");

		BiddingBaseManager manager = biddingManagerFactory
			.createManager(mock(IUnityAdsTokenListener.class), experiments);
		assertThat(manager, instanceOf(BiddingLazyManager.class));
	}

	@Test
	public void createManagerExperimentHybrid() {
		when(experiments.getScarBiddingManager()).thenReturn("hyb");

		BiddingBaseManager manager = biddingManagerFactory
			.createManager(mock(IUnityAdsTokenListener.class), experiments);
		assertThat(manager, instanceOf(BiddingOnDemandManager.class));
	}

	@Test
	public void createManagerExperimentDisabled() {
		when(experiments.getScarBiddingManager()).thenReturn("dis");

		BiddingBaseManager manager = biddingManagerFactory
			.createManager(mock(IUnityAdsTokenListener.class), experiments);
		assertThat(manager, instanceOf(BiddingDisabledManager.class));
	}
}

File: unity-ads/src/test/java/com/unity3d/services/ads/gmascar/managers/BiddingManagerFactoryTest.kt
package com.unity3d.services.ads.gmascar.managers

import com.unity3d.ads.IUnityAdsTokenListener
import com.unity3d.services.ads.gmascar.GMA
import com.unity3d.services.core.configuration.IExperiments
import io.mockk.*
import io.mockk.impl.annotations.MockK
import org.hamcrest.CoreMatchers.instanceOf
import org.hamcrest.MatcherAssert.assertThat
import kotlin.test.AfterTest
import kotlin.test.BeforeTest
import kotlin.test.Test

class BiddingManagerFactoryTest {

    private lateinit var biddingManagerFactory: BiddingManagerFactory

    @MockK
    private lateinit var experiments: IExperiments

    @MockK
    private lateinit var gma: GMA

    @BeforeTest
    fun setup() {
        biddingManagerFactory = BiddingManagerFactory.getInstance()

        MockKAnnotations.init(this, relaxed = true)

        mockkStatic(GMA::class)
        every { GMA.getInstance() } returns gma
    }

    @AfterTest
    fun tearDown() {
        unmockkStatic(GMA::class)
    }

    @Test
    fun createManagerScarSupportedNullExperiments() {
        every { gma.hasSCARBiddingSupport() } returns true
        val manager = biddingManagerFactory.createManager(IUnityAdsTokenListener { }, null)
        assertThat(manager, instanceOf(BiddingEagerManager::class.java))
    }

    @Test
    fun createManagerScarUnsupportedNullExperiments() {
        every { gma.hasSCARBiddingSupport() } returns false
        val manager = biddingManagerFactory.createManager(IUnityAdsTokenListener { }, null)
        assertThat(manager, instanceOf(BiddingDisabledManager::class.java))
    }

    @Test
    fun createManagerScarUnsupportedValidExperiments() {
        every { gma.hasSCARBiddingSupport() } returns false
        val manager = biddingManagerFactory.createManager(IUnityAdsTokenListener { }, experiments)
        assertThat(manager, instanceOf(BiddingDisabledManager::class.java))
    }

    @Test
    fun createManagerScarSupportedExperimentTrue() {
        every { gma.hasSCARBiddingSupport() } returns true
        every { experiments.isScarBannerHbEnabled } returns true

        val manager = biddingManagerFactory.createManager(IUnityAdsTokenListener { }, experiments)
        assertThat(manager, instanceOf(BiddingEagerManager::class.java))
    }

    @Test
    fun createManagerScarSupportedExperimentFalse() {
        every { gma.hasSCARBiddingSupport() } returns true
        every { experiments.isScarBannerHbEnabled } returns false

        val manager = biddingManagerFactory.createManager(IUnityAdsTokenListener { }, experiments)
        assertThat(manager, instanceOf(BiddingEagerManager::class.java))
    }
}